src/core/async-work-cache.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/**
+ * A map from keys to promises of values. Used for caching asynchronous work.
+ */
+export class AsyncWorkCache<K, V> {
+  private _keyToResultMap: Map<K, Promise<V>>;
+  constructor(from?: AsyncWorkCache<K, V>) {
+    if (from) {
+      this._keyToResultMap = new Map(from._keyToResultMap);
+    } else {
+      this._keyToResultMap = new Map();
+    }
+  }
+
+  /**
+   * If work has already begun to compute the given key, return a promise for
+   * the result of that work.
+   *
+   * If not, compute it with the given function.
+   *
+   * This method ensures that we will only try to compute the value for `key`
+   * once, no matter how often or with what timing getOrCompute is called, even
+   * recursively.
+   */
+  async getOrCompute(key: K, compute: () => Promise<V>) {
+    const cachedResult = this._keyToResultMap.get(key);
+    if (cachedResult) {
+      return cachedResult;
+    }
+    const promise = (async() => {
```

Comment:
Yes, but I prefer the more explicit style here, as I also want to leave in the comment about why we have the `Promise.resolve()`